### ISSUE TITLE
Allow exposing optical media devices

### ISSUE THREAD

**Issue by alexlarsson** [flatpak/flatpak#12]
_From @hadess on April 22, 2016 10:5_

For DVD and VCD playback to work in totem, the application might need access to the device itself, as well as the mount point for that filesystem (to check what is on the disc).

_Copied from original issue: alexlarsson/xdg-app#139_

**Comment by alexlarsson** [comment]
I understand the need, but direct block device access is scary stuff it things go wrong. I mean, a cdrom device is just a block device node like any other, such as your ide disks...

How does user permissions for dvd devices work? What makes it possible for the user to read the dvd drive device?

**Comment by alexlarsson** [comment]
_From @hadess on April 22, 2016 11:47_

Right now? systemd-udev has rules to mark "CD drives" (meaning any optical media) as such:
/usr/lib/udev/rules.d/50-udev-default.rules
and
/usr/lib/udev/rules.d/60-cdrom_id.rules

I don't know what changes the permissions to access it though.

**Comment by magcius** [comment]
Read-only media doesn't seem too dangerous... is it?

**Comment by alexlarsson** [comment]
I don't mean that read only access to cdrom is all that dangerous, but that exposing block devices in general is a tricky area to get right. There is nothing "special" about cdrom devices, they are just some random /dev/sd* name, and you have to do i/o on the device node to know what type it is, and this can change dynamically over time.

**Comment by magcius** [comment]
Hm, I thought there was /dev/cdrom and such. Could we link in /dev/by-id/etc. and ask people to just use udev?

**Comment by hadess** [comment]
The filename of the device isn't the problem. There's a problem with discovery, and hotplugging of the device, and ensuring that the device is only available to applications when specific discs are inside.

Eg. playback application should be able to access the data on the disc, and poke the device (via the device node) only if there's a DVD, or VCD inside.

**Comment by hadess** [comment]
I forgot the second part of the question: udev client libraries aren't in the platform because the protocol used by udev isn't considered stable.

We'll need discovery methods similar to PulseAudio, pinos, and others for optical discs (though it could be implemented in gvfs, see https://bugzilla.gnome.org/show_bug.cgi?id=766914)

**Comment by alexlarsson** [comment]
We discussed this a bit at the hackfest. One option is to have a permission option to have full /dev access. This is not something that you should normally use, but it definately has its place if you're mostly using flatpak for distribution, not sandboxing. I will add this to flatpak.

We also have some ideas of using fuse to allow dynamic access to some limited subset of device types. I will have a look at how doable this is.

**Comment by alexlarsson** [comment]
@hadess Does totem need to do anything but reads/seeks on the block device? I.e. does it need any ioctls or similar weird things?

**Comment by hadess** [comment]
It does need a number of ioctls for DVDs to work. I don't know what would be necessary to play back VCDs or Blu-Ray, but I imagine that there are equivalent calls for those.

**Comment by alexlarsson** [comment]
For now, --device=all which gives full access to the host /dev.

**Comment by victorlevasseur** [comment]
I'm trying to bundle SFML with flatpak but it can't find the udev libraries in the org.Gnome.Sdk SDK. Will the udev libs be ever included in the platform or not ?

**Comment by hadess** [comment]
> I'm trying to bundle SFML with flatpak but it can't find the udev libraries in the org.Gnome.Sdk SDK. Will the udev libs be ever included in the platform or not ?

No, udev's API/ABI isn't stable, so udev libraries will never be available in the Sdk, as you don't want to rely on a particular version of the udev daemon in the host.

That's why there are bugs opened about providing a way to discover things like optical media drives and joysticks. If you can think of specific device classes that will need to be accessed by applications, please file a separate bug.

**Comment by victorlevasseur** [comment]
Ok, thanks for your answer. I was just asking because SFML depends on the udev libraries :(

**Comment by hadess** [comment]
SDL at least has the ability to disable the udev dependency. Does SFML not have the same option?

**Comment by victorlevasseur** [comment]
No, I'm currently trying to write a patch (to disable the joystick support first).

**Comment by Jehan** [comment]
GIMP has an optional dependency to libgudev, which in turns obviously depends on libudev. Of course, we can just deactivate the option, but we'd like to have the official GIMP flatpak as complete as possible, obviously for best experience. What is the current proposed approach?

I've tried to build libudev, which basically means build the whole systemd into flatpak! (as of now, I could not find any option in systemd configure script to only build udev) That seems overkill. I still tried, but the build failed (for a reason I have not fully investigated yet, because the configure part was OK so conditions seemed met).

Anyway I thought it was worth discussing our need here too.

**Comment by hadess** [comment]
> GIMP has an optional dependency to libgudev, which in turns obviously depends on libudev. Of course, we can just deactivate the option, but we'd like to have the official GIMP flatpak as complete as possible, obviously for best experience. What is the current proposed approach?

None. Don't try to compile libudev, it won't work inside the sandbox. You'll need to figure out why libudev is being used, and replace each of those uses with portals. I doubt that GIMP is using libudev to manage its own storage media, so this isn't the right issue to discuss this.

**Comment by alexlarsson** [comment]
Also, libudev is not accidentally left out of the runtime, I asked upstream about the stability guarantees, and it turns out that udevd (which writes the files libudev reads) does not guarantee in-file backwards compat, only libudev ABI. This means we can't put a libudev in the sandbox as it may not match what is running on the host.

Also, like @hadess  said, in general udev is for hardware-stuff, drivers and whatnot. I'm not sure why gimp would need it at all.

**Comment by alexlarsson** [comment]
It would be interesting to know if you did any research on this what exactly it ends up using it for.

**Comment by hadess** [comment]
> It would be interesting to know if you did any research on this what exactly it ends up using it for.

Pretty sure it uses it for input devices that it accesses directly (Edit -> Preferences -> Input Devices -> Input Controllers -> Linux Input).

**Comment by alexlarsson** [comment]
Similar to https://github.com/flatpak/flatpak/issues/7 then

**Comment by hadess** [comment]
> Similar to #7 then

A bit broader, but yes, that's it. Joystick could be replaced there by "too complicated/too rare an input device to be handled by the Wayland/GTK+ input APIs"

**Comment by Jehan** [comment]
> It would be interesting to know if you did any research on this what exactly it ends up using it for.

Yes I did investigate. We only use libgudev (i.e. libudev) in a single file (`modules/gimpinputdevicestore-gudev.c`), a module for input devices. It seems it would be used for special kind of devices, like midi stuff, because GIMP has such a feature to be controlled through all kind of funny devices.

To be clear, that's something I have never used myself and this is definitely not a very commonly used feature, I think. So I've discussed with Mitch and we'll just disable this in the flatpak build.

I will maybe make a bug report on our side to have this replaced by an alternative implementation (though I would not be astonished such a re-implementation never comes, especially since the workaround is just to not use flatpak).
Do you have any knowledge of which lib can be used to get similar feature as libudev while still flatpak-friendly? There is no need for someone to work on this feature if it ends up the same in the end, and not working in a flatpak.

**Comment by hadess** [comment]
> Do you have any knowledge of which lib can be used to get similar feature as libudev while still flatpak-friendly? There is no need for someone to work on this feature if it ends up the same in the end, and not working in a flatpak.

Not yet, this would need a portal. Help implementing it, or at least drawing the API would be appreciated.

**Comment by Jehan** [comment]
> Not yet, this would need a portal. Help implementing it, or at least drawing the API would be appreciated.

Ok message noted. :-)
I can't promise anything though, because that's not a top priority for me either. But if I can make some time at some point, I will help.

Also I would suggest to maybe write some doc on the relationship between portals and permissions. From your main developer page, the only mention to portals is in the top graphics (section "Key Concept"). That makes this all a little fuzzy. :-)

**Comment by alexlarsson** [comment]
It sounds fuzzy because it is a bit fuzzy. We're somewhat building things as we go, and thing show up, and "portal" is a generic word to mean "some service that safely lets you extend what sandboxed apps can do".

**Comment by Jehan** [comment]
Raising this back from the dead since by mindelessly wandering on the `shared-modules` repo of flathub today, I discover: https://github.com/flathub/shared-modules/tree/master/udev

Is this to say that it is now possible to build libudev in a flatpak?

**Comment by hadess** [comment]
> Is this to say that it is now possible to build libudev in a flatpak?

It's an old version of udev, just for the purpose of shutting up builds that require libudev. It doesn't actually work for enumerating devices, or using hotplug.

**Comment by hadess** [comment]
Updated gvfs RFE, for device enumeration:
https://gitlab.gnome.org/GNOME/gvfs/issues/281
